wurst/systems/entities/bosses/BossUtilities.wurst

```diff
@@ -0,0 +1,54 @@
+package BossUtilities
+
+// Standard-library imports:
+import Assets
+import ClosureTimers
+
+// Local imports:
+import HealingSystem
+import GroupUtils
+import UnitEntity
+
+public function checkInvulnerability(region bossRegion, unit boss) returns bool
+    // The Boss is always invulnerable when outside its arena.
+    if not boss.isInRegion(bossRegion)
+        return true
+
+    // The Boss is invulnerable unless there is a troll present in its arena.
+    for troll in getTrolls()
+        // Disable invulnerability if both conditions are met.
+        if troll.isAlive() and troll.isInRegion(bossRegion)
+            return false
+
+    // Enable invulnerability if no matches were found.
+    return true
+
+public function toggleInvulnerability(region bossRegion, unit boss)
+    // Exit if The Boss is not currently alive.
+    if not boss.isAlive()
+        return
+
+    // Toggle the invulnerability.
+    if checkInvulnerability(bossRegion, boss)
+        boss.addAbility(AbilityIds.invulnerable)
+        toggleHealing(boss)
+    else
+        boss.removeAbility(AbilityIds.invulnerable)
+
+// TODO: rework this into an ability
+function toggleHealing(unit boss)
+    let hpRegen = 4.
+    let regenPeriod = 1.
+    
+    // Heal the boss 4 hp every 1 second
+    doPeriodically(regenPeriod) (CallbackPeriodic cb) ->
+        // Use healing system
+        new HealingInstance(boss, hpRegen, HealingType.OTHER)
+        // If boss is not invulnerable or is max health
+        if not boss.isInvulnerable() or boss.getMissingHP() == 0
+            // Stop healing
+            destroy cb
+
+public function checkEntity(UnitEntity bossEntity, region bossRegion, unit boss)
```

Comment:
You don't need to pass `boss` directly. Just use `bossEntity.getUnit()`. Once you don't need that identifier, you can use `boss` instead of `bossEntity` as the entity is implied by the type.

Also, you should be able to name this function `toggleInvulnerability` because Wurst allows function overloading.